### DISCUSSION TITLE
Image.openslideload only accesses the first tile in Mirax (MRXS) files.

### DISCUSSION THREAD

**Discussion by mth-gsk** [libvips/libvips#2062] [Q&A]
Attempting to process a Mirax (MRXS) histopathology image and need to save it in a format compatible with commercial slide viewers (should be pyramid TIFF). 

Observed with Openslide test data (Mirax2.2-1.zip) from [here](http://openslide.cs.cmu.edu/download/openslide-testdata/Mirax/) as well as new data straight from the scanner. 

Minimal code to reproduce is as follows. 
```
import libvips

mirax_path = path_to_your_mirax_file
img = pyvips.Image.openslideload(mirax_path, level=4)
img.tiffsave('out.tiff', tile=True, compression='jpeg', Q=90, pyramid=True)
```
using `level=4` to accelerate the processing. 

Expected a pyramid TIFF representing the scan of the tissue slide. 

Resulted in only one tile worth of data (white background) transcribed into the TIFF. Varying the level increases the observed size of the patch in the top left corner, seeming consistent with an issue with obtaining tile data. Screenshots attached below. 

<img width="517" alt="Screenshot 2020-10-20 at 13 17 26" src="https://user-images.githubusercontent.com/33751992/96595674-d710a300-12e3-11eb-8921-c15ae37870b2.png">
<img width="510" alt="Screenshot 2020-10-20 at 13 17 18" src="https://user-images.githubusercontent.com/33751992/96595681-d8da6680-12e3-11eb-9d38-3cf5b43102b0.png">

**Environment**
(please complete the following information)
 - OS: MacOS 10.15.4
 - Vips: 8.10
- pyvips: 2.1.8
- openslide: 1.1.2

**Comment by jcupitt** [comment]
Hello @mth-gsk,

Sorry, I think one of the libraries you are using probably has a bug. pixman has had problems recently, so I would try other versions of that, or other versions of cairo.

I tried Ubuntu 20.04 and that fails too, though in a different way (due to pixman, I think). I tried on Ubuntu 18.04 and that worked fine:

https://github.com/jcupitt/docker-builds/tree/master/pyvips-libvips-openslide-ubuntu18.04

It'll need some investigation to track down.

**Comment by jcupitt** [comment]
I removed the bug label since this is probably a problem with your library stack rather than with libvips.

**Comment by jcupitt** [comment]
I tried rebuilding against pixman 0.40 (current stable) and that's enough to make it work on Ubuntu 20.04. Out of the box, ubuntu 20.04 is still using pixman 0.38.4, which is known to be broken.

```
vips copy Mirax2.2-1.mrxs[level=8] x.tif[tile,pyramid,compression=jpeg,Q=90]
vips copy x.tif x.jpg
```

Makes:

![x](https://user-images.githubusercontent.com/580843/96604586-30310480-12ed-11eb-8ce9-b46b1f54aca8.jpg)

That's libvips 8.10, pyvips 2.1.12, openslide 3.4.1, libcairo 1.16.0-4.

**Comment by mth-gsk** [comment]
@jcupitt ok, I will have a look at the pixman version

**Comment by mth-gsk** [comment]
@jcupitt I have updated pyvips to 2.1.12 and my pixman was at 0.40 already. I have same versions of cairo and openslide as you listed above -I am still seeing the same behaviour on my Mac.

**Comment by jcupitt** [comment]
Wait, could it be a bug in Apple's Preview?

I tried:

```
vips copy Mirax2.2-1.mrxs[level=8] x.tif[tile,pyramid,compression=jpeg,Q=90]
vips copy x.tif x.jpg
open x.jpg
```

On my Mac and it seems fine.

Do you have another image view program you could test with?

**Comment by jcupitt** [comment]
nip2 on Mac can view `x.tif` correctly:

https://github.com/libvips/nip2/releases/tag/v8.6.0-beta2

There's a DMG there you can download.

**Comment by mth-gsk** [comment]
Tested with QuPath, same behaviour as with Apple Preview when accessing the tiff.
When converted to jpeg, I get the broad view but numerous artefacts are present (on internal data, will check with the public one).

@jcupitt similar issues seen with the publicly available MRXS:
![test](https://user-images.githubusercontent.com/33751992/96990410-fc3d2580-151e-11eb-940e-1edbf987d993.jpg)

**Comment by jcupitt** [comment]
I had images that looked a bit like that on ubuntu 20.04 until I updated pixman. It seems to be working fine there now.

I just tried this on my mac, running catalina 10.15.7, using your image:

```
12:53 $ brew update && brew upgrade
Already up-to-date.
✔ ~/pics/openslide 
12:55 $ vips copy Mirax2.2-1.mrxs[level=4] x.tif[pyramid,tile,compression=jpeg]
✔ ~/pics/openslide 
12:55 $ vips copy x.tif x.jpg
✔ ~/pics/openslide 
```

To make:

![x](https://user-images.githubusercontent.com/580843/97000940-6d84d480-152f-11eb-8ac2-485d517eaa17.jpg)

So I think it's all working for me.

**Comment by jcupitt** [comment]
Here's how my vips exe is linked:

```
12:59 $ otool -L $(which vips)
/usr/local/bin/vips:
	/usr/local/Cellar/vips/8.10.2/lib/libvips.42.dylib (compatibility version 55.0.0, current version 55.4.0)
	/usr/local/opt/libheif/lib/libheif.1.dylib (compatibility version 11.0.0, current version 11.1.0)
	/usr/local/opt/imagemagick/lib/libMagickCore-7.Q16HDRI.7.dylib (compatibility version 8.0.0, current version 8.0.0)
	/usr/local/opt/libomp/lib/libomp.dylib (compatibility version 5.0.0, current version 5.0.0)
	/usr/local/opt/liblqr/lib/liblqr-1.0.dylib (compatibility version 4.0.0, current version 4.2.0)
	/usr/local/opt/xz/lib/liblzma.5.dylib (compatibility version 8.0.0, current version 8.5.0)
	/usr/lib/libbz2.1.0.dylib (compatibility version 1.0.0, current version 1.0.5)
	/usr/local/opt/libtool/lib/libltdl.7.dylib (compatibility version 11.0.0, current version 11.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
	/usr/local/opt/libspng/lib/libspng.0.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/libpng/lib/libpng16.16.dylib (compatibility version 54.0.0, current version 54.0.0)
	/usr/local/opt/libimagequant/lib/libimagequant.0.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/libtiff/lib/libtiff.5.dylib (compatibility version 11.0.0, current version 11.0.0)
	/usr/local/opt/mozjpeg/lib/libjpeg.8.dylib (compatibility version 10.0.0, current version 10.2.0)
	/usr/local/opt/glib/lib/libgmodule-2.0.0.dylib (compatibility version 6601.0.0, current version 6601.1.0)
	/usr/lib/libexpat.1.dylib (compatibility version 7.0.0, current version 8.0.0)
	/usr/local/opt/pango/lib/libpangoft2-1.0.0.dylib (compatibility version 4601.0.0, current version 4601.2.0)
	/usr/local/opt/pango/lib/libpango-1.0.0.dylib (compatibility version 4601.0.0, current version 4601.2.0)
	/usr/local/opt/harfbuzz/lib/libharfbuzz.0.dylib (compatibility version 20702.0.0, current version 20702.0.0)
	/usr/local/opt/fontconfig/lib/libfontconfig.1.dylib (compatibility version 14.0.0, current version 14.0.0)
	/usr/local/opt/freetype/lib/libfreetype.6.dylib (compatibility version 24.0.0, current version 24.3.0)
	/usr/local/opt/libgsf/lib/libgsf-1.114.dylib (compatibility version 115.0.0, current version 115.47.0)
	/usr/local/opt/libxml2/lib/libxml2.2.dylib (compatibility version 12.0.0, current version 12.10.0)
	/usr/local/opt/fftw/lib/libfftw3.3.dylib (compatibility version 9.0.0, current version 9.8.0)
	/usr/local/opt/orc/lib/liborc-0.4.0.dylib (compatibility version 33.0.0, current version 33.0.0)
	/usr/local/opt/little-cms2/lib/liblcms2.2.dylib (compatibility version 3.0.0, current version 3.10.0)
	/usr/local/opt/giflib/lib/libgif.dylib (compatibility version 0.0.0, current version 7.2.0)
	/usr/local/opt/librsvg/lib/librsvg-2.2.dylib (compatibility version 50.0.0, current version 50.0.0)
	/usr/local/opt/glib/lib/libgio-2.0.0.dylib (compatibility version 6601.0.0, current version 6601.1.0)
	/usr/local/opt/gdk-pixbuf/lib/libgdk_pixbuf-2.0.0.dylib (compatibility version 4001.0.0, current version 4001.0.0)
	/usr/local/opt/poppler/lib/libpoppler-glib.8.dylib (compatibility version 8.0.0, current version 8.19.0)
	/usr/local/opt/glib/lib/libgobject-2.0.0.dylib (compatibility version 6601.0.0, current version 6601.1.0)
	/usr/local/opt/glib/lib/libglib-2.0.0.dylib (compatibility version 6601.0.0, current version 6601.1.0)
	/usr/local/opt/gettext/lib/libintl.8.dylib (compatibility version 11.0.0, current version 11.0.0)
	/usr/local/opt/cairo/lib/libcairo.2.dylib (compatibility version 11603.0.0, current version 11603.0.0)
	/usr/local/opt/openexr/lib/libIlmImf-2_5.25.dylib (compatibility version 25.0.0, current version 25.0.2)
	/usr/local/opt/ilmbase/lib/libImath-2_5.25.dylib (compatibility version 25.0.0, current version 25.0.2)
	/usr/local/opt/ilmbase/lib/libHalf-2_5.25.dylib (compatibility version 25.0.0, current version 25.0.2)
	/usr/local/opt/ilmbase/lib/libIex-2_5.25.dylib (compatibility version 25.0.0, current version 25.0.2)
	/usr/local/opt/ilmbase/lib/libIexMath-2_5.25.dylib (compatibility version 25.0.0, current version 25.0.2)
	/usr/local/opt/ilmbase/lib/libIlmThread-2_5.25.dylib (compatibility version 25.0.0, current version 25.0.2)
	/usr/local/opt/openslide/lib/libopenslide.0.dylib (compatibility version 5.0.0, current version 5.1.0)
	/usr/local/opt/cfitsio/lib/libcfitsio.9.dylib (compatibility version 9.0.0, current version 9.3.49)
	/usr/local/opt/webp/lib/libwebpmux.3.dylib (compatibility version 4.0.0, current version 4.5.0)
	/usr/local/opt/webp/lib/libwebpdemux.2.dylib (compatibility version 3.0.0, current version 3.6.0)
	/usr/local/opt/webp/lib/libwebp.7.dylib (compatibility version 9.0.0, current version 9.0.0)
	/usr/local/opt/libmatio/lib/libmatio.9.dylib (compatibility version 11.0.0, current version 11.2.0)
	/usr/local/opt/hdf5/lib/libhdf5.200.dylib (compatibility version 201.0.0, current version 201.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/local/opt/libexif/lib/libexif.12.dylib (compatibility version 16.0.0, current version 16.4.0)
```

So it's just a straight homebrew build.

What commands are you running exactly? Could you be using old versions of some library in some other part of your system?

**Comment by jcupitt** [comment]
I tried in QuPath and you're right, it just shows one tile there too. This is a bug in QuPath and you should probably report it.

Try (for example):

```
vips copy IMG_0513.jpg x.tif[tile,compression=jpeg]
```

With any image and you'll see the same problem -- QuPath does not seem to support tiled, compressed tiff. 

Perhaps it is using the Apple libtiff? That would explain why it fails on the same images as Preview.

**Comment by jcupitt** [comment]
To prove it's a bug in QuPath and Preview, try this:

```
13:22 $ vips copy IMG_0513.jpg x.tif[tile,compression=jpeg]
✔ ~/pics 
13:23 $ tiffcp -s -c none x.tif y.tif
✔ ~/pics 
13:23 $ open y.tif
✔ ~/pics 
13:23 $ 
```

ie. make a tiled, compressed TIFF with libvips, then use tiffcp (the definitive TIFF reader) to convert from that to compressed strip TIFF before passing it to Preview. You shuld find it works.

**Comment by jcupitt** [comment]
(that's hte homebrew tiffcp, by the way)

**Comment by jcupitt** [comment]
Bug summary:

Here's a demo image:

http://www.rollthepotato.net/~john/tiled-jpeg-tiff-pyramid.tif

Made with:

```
vips copy Mirax2.2-1.mrxs[level=4] tiled-jpeg-tiff-pyramid.tif[tile,pyramid,compression=jpeg]
```

It works fine on Ubuntu 20.04 and Windows 10 using the image preview, Photoshop, Gimp, paint.net and nip2.

It does not work on macos using either Preview or QuPath.

You can convert it to an uncompressed tiff on macos using homebrew `tiffcp`, at which point Preview and QuPath will display it correctly.

**Comment by jcupitt** [comment]
I opened a bug report on QuPath, fyi.

**Comment by jcupitt** [comment]
Heh so that was a dead end, QuPath seems fine.

The problem seems to be in homebrew's libtiff -- it's a straight build of libtiff 4.1 and has a series of known bugs around JPEG-in-TIFF. Ubuntu also use libtiff 4.1, but they apply a set of patches to fix this.

Here's a simple demo image:

http://www.rollthepotato.net/~john/broken-macos-tiff.tif

Created on macOS like this:

```
vips copy img.jpg x.tif[tile,compression=jpeg]
```

You can view it on Ubuntu, but macOS preview fails (just shows the top-left tile).

To fix this, I think someone needs to revise the libtiff formula in homebrew to include the set of patches that Ubuntu apply.
